File: CCIP/CCIP/UIAlertController+additional.m
//
//  UIAlertController+Labels.m
//  iOS Portal
//
//  Created by 腹黒い茶 on 25/2/2015.
//  Copyright (c) 2015年 KFSYSCC. All rights reserved.
//

#import "UIAlertController+additional.h"

@implementation UIApplication (additional)

+ (UIViewController *)getMostTopPresentedViewController {
    UIViewController *vc = [[[UIApplication sharedApplication] keyWindow] rootViewController];
    while ([vc presentedViewController])
        vc = [vc presentedViewController];
    return vc;
}

@end

@implementation UIAlertController (additional)
@dynamic titleLabel;
@dynamic messageLabel;

- (NSArray *)viewArray:(UIView *)root {
    //NSLog(@"%@", root.subviews);
    static NSArray *_subviews = nil;
    _subviews = nil;
    for (UIView *v in root.subviews) {
        if (_subviews) {
            break;
        }
        if ([v isKindOfClass:[UILabel class]]) {
            _subviews = root.subviews;
            return _subviews;
        }
        [self viewArray:v];
    }
    return _subviews;
}

- (UILabel *)titleLabel {
    return [self viewArray:self.view][0];
}

- (UILabel *)messageLabel {
    return [self viewArray:self.view][1];
}

+ (UIAlertController *)actionSheet:(id)sender withTitle:(NSString *)title andMessage:(NSString *)message {
    UIAlertController *ac = [self alertControllerWithTitle:title
                                                   message:message
                                            preferredStyle:UIAlertControllerStyleActionSheet];
    UIView *sd = sender;
    CGRect frame = [sd frame];
    frame.origin.x += frame.size.width / 2.0f;
    frame.origin.y += frame.size.height / 2.0f;
    frame.size.width = 1.0f;
    frame.size.height = 1.0f;
    sd = [sd superview];
    while(![[[sd class] description] hasSuffix:@"ViewController"] && sd != nil) {
        CGRect f = [sd frame];
        sd = [sd superview];
        frame.origin.x += f.origin.x;
        frame.origin.y += f.origin.y;
    }
    [ac.popoverPresentationController setSourceView:[[UIApplication getMostTopPresentedViewController] view]];
    [ac.popoverPresentationController setSourceRect:frame];
    return ac;
}

+ (UIAlertController *)alertOfTitle:(NSString *)title
                        withMessage:(NSString *)message
                   cancelButtonText:(NSString *)cancelText
                        cancelStyle:(UIAlertActionStyle)cancelStyle
                       cancelAction:(void (^)(UIAlertAction *))cancelAction {
    UIAlertController *ac = [UIAlertController alertControllerWithTitle:title
                                                                message:message
                                                         preferredStyle:UIAlertControllerStyleAlert];
    [ac setModalPresentationStyle:UIModalPresentationPopover];
    [ac addActionButton:cancelText
                  style:cancelStyle
                handler:cancelAction];
    return ac;
}

- (void)showAlert:(void (^)(void))completion {
    [[UIApplication getMostTopPresentedViewController] presentViewController:self
                                                                    animated:YES
                                                                  completion:completion];
}

- (void)addActionButton:(NSString *)title
                  style:(UIAlertActionStyle)style
                handler:(void (^)(UIAlertAction *))handler {
    UIAlertAction *action = [UIAlertAction actionWithTitle:title
                                                     style:style
                                                   handler:handler];
    [self addAction:action];
}

@end

File: OPass/Class+addition/UIAlertController+additional.swift
//
//  UIAlertController+additional.swift
//  OPass
//
//  Created by 腹黒い茶 on 2018/10/7.
//  2018 OPass.
//

import Foundation
import UIKit

@objc extension UIAlertController {
    var titleLabel: UILabel {
        get {
            if let lb = self.viewArray(self.view)[0] as? UILabel {
                return lb
            }
            return UILabel.init()
        }
    }
    var messageLabel: UILabel {
        get {
            if let lb = self.viewArray(self.view)[1] as? UILabel {
                return lb
            }
            return UILabel.init()
        }
    }
    func viewArray(_ root: UIView) -> NSArray {
        let _subviews: NSArray? = nil
        for v: UIView in root.subviews {
            if (_subviews != nil) {
                break
            }
            if (v.isKind(of: UILabel.self)) {
                return root.subviews as NSArray
            }
            return self.viewArray(v)
        }
        if let sv = _subviews {
            return sv
        }
        return []
    }

    static func actionSheet(
        _ sender: Any,
        withTitle: String,
        andMessage: String
    ) -> UIAlertController {
        let ac: UIAlertController = self.init(title: withTitle, message: andMessage, preferredStyle: UIAlertController.Style.actionSheet)
        guard var sd = sender as? UIView else { return ac }
        var frame: CGRect = sd.frame
        frame.origin.x += frame.size.width / 2.0
        frame.origin.y += frame.size.height / 2.0
        frame.size.width = 1.0
        frame.size.height = 1.0
        if let _sd = sd.superview { sd = _sd }
        while (!sd.self.description.hasSuffix("ViewController")) {
            let f: CGRect = sd.frame
            if let _sd = sd.superview { sd = _sd }
            frame.origin.x += f.origin.x
            frame.origin.y += f.origin.y
        }
        ac.popoverPresentationController?.sourceView = UIApplication.getMostTopPresentedViewController()?.view
        ac.popoverPresentationController?.sourceRect = frame
        return ac
    }
    @objc static func alertOfTitle(
        _ title: String?,
        withMessage: String?,
        cancelButtonText: String,
        cancelStyle: UIAlertAction.Style,
        cancelAction: ((UIAlertAction) -> Void)?
    ) -> UIAlertController {
        let ac: UIAlertController = self.init(title: title, message: withMessage, preferredStyle: UIAlertController.Style.alert)
        ac.modalPresentationStyle = UIModalPresentationStyle.popover
        ac.addActionButton(cancelButtonText, style: cancelStyle, handler: cancelAction)
        return ac
    }
    func showAlert(_ completion: @escaping () -> Void) {
        UIApplication.getMostTopPresentedViewController()?.present(self, animated: true, completion: completion)
    }
    func addActionButton(
        _ title: String,
        style: UIAlertAction.Style,
        handler: ((UIAlertAction) -> Void)?
    ) {
        let action: UIAlertAction = UIAlertAction.init(title: title, style: style, handler: handler)
        self.addAction(action)
    }
}
